### ISSUE TITLE
VK_PRESENT_MODE_IMMEDIATE_KHR locked to display refresh rate on macOS

### ISSUE THREAD

**Issue by SRSaunders** [KhronosGroup/MoltenVK#1925]
Using mode `VK_PRESENT_MODE_IMMEDIATE_KHR` I expect no-delay presentation (i.e. tearing) with an unlocked frame rate.  Instead I observe a frame rate locked to my display refresh rate (60 Hz).  Using the Optick profiler, it appears the presentation engine is still using all swapchain buffers and waiting for vsync, which is counter to the spec.

In contrast, the same code behaves properly on Windows and Linux - i.e. immediate presentation with an unlocked frame rate.

Is there something special about the setup requirements for this presentation mode on macOS with MoltenVK?  Note that I am using triple buffering (3 GPU frames + 3 swapchain buffers) with `synchronousQueueSubmits = VK_FALSE`.  Using the Optick profiler, I can see the application is waiting on GPU graphics queue completion and not on acquire, submit or present, which is a bit puzzling since everything is idle (CPU + GPU) for a long period of time between vsync events.

This may be a duplicate of #581 (which is still open with no apparent resolution).

**Comment by billhollings** [comment]
`VK_PRESENT_MODE_IMMEDIATE_KHR` should be working okay.

PR #1927 enhances the macOS Cube demo to permit enabling `VK_PRESENT_MODE_IMMEDIATE_KHR` in the `macOS/ DemoViewController.m` file. Try that out and see if it behaves as you expect. If so, we'll need to figure out how your app is using it differently than the demo. There may be some subtle secondary interactions that are showing up.

**Comment by SRSaunders** [comment]
Thanks for updating the Cube demo.  I will try it out asap as reference point.

However, in the mean time I have been doing some digging and think I have found a trigger for the issue, but not the underlying cause.  As soon as I define an instance of `vk::PresentTimesInfoGOOGLE` in the pNext chain of `Queue::presentKHR()`, the problem arises and **VK_PRESENT_MODE_IMMEDIATE_KHR** locks to my display refresh rate.  When it is not included in the pNext chain, FPS is unlocked and present behaves as I would expect.

I don't know if this is intrinsic to the way the feature works, but it does correspond to my observed platform differences: on Windows and Linux (no FPS locking) the feature is not available for my AMD card and thus is not included in the pNext chain, however it is with MoltenVK on macOS where FPS locking is evident.

Strange coincidence given our recent interaction on improving the present times feature :)

**Comment by billhollings** [comment]
Thanks for all the research.

If `VkPresentTimesInfoGOOGLE` and `VkPresentTimeGOOGLE` are included, then `VkPresentTimeGOOGLE::desiredPresentTime` sets the time at which the swapchain image will be presented, and MoltenVK calls the Metal operation [`presentAtTime:`](https://developer.apple.com/documentation/metal/mtldrawable/1470282-presentattime?language=objc).

So the incorrect display sync could be the result of either the app setting `desiredPresentTime` to a value that matches platform display refresh, or Metal's `presentAtTime:` syncing to display refresh under the covers, unexpectedly, and without documentation to that effect.

Can you determine which of those are occurring in your app?

**Comment by SRSaunders** [comment]
It's the latter, since my `desiredPresentTime = 0` in all cases.

The following code inside `MVKPresentableSwapchainImage::presentCAMetalDrawable()` fixes it.  However, I'm not sure if this is how you want to solve it.  For instance, it appears the meaning of `presentInfo.hasPresentTime` may be overloaded, and currently indicates that both a presentedHandler and desiredPresentTime are requested.  In my case I am only interested in the presented time handler, and my desired present time is not defined (i.e. = 0).

```
if (presentInfo.hasPresentTime) {
	// Convert from nsecs to seconds for Metal
	addPresentedHandler(mtlDrwbl, presentInfo);
	if (presentInfo.desiredPresentTime) {
		[mtlDrwbl presentAtTime: (double)presentInfo.desiredPresentTime * 1.0e-9];
	}
	else {
		[mtlDrwbl present];
	}
} else {
	[mtlDrwbl present];
}
```

During debugging, I also observe that `presentInfo.presentMode = VK_PRESENT_MODE_MAX_ENUM_KHR`. This causes the code to skip setting `mtlDrwbl.layer.displaySyncEnabledMVK`. Is this a red herring, or something to be concerned about?

**Comment by cdavis5e** [comment]
> During debugging, I also observe that `presentInfo.presentMode = VK_PRESENT_MODE_MAX_ENUM_KHR`. This causes the code to skip setting `mtlDrwbl.layer.displaySyncEnabledMVK`. Is this a red herring, or something to be concerned about?

Probably a red herring. According to the [docs](https://developer.apple.com/documentation/quartzcore/cametallayer/2887087-displaysyncenabled?language=objc):

> The default value is `YES`.

But that actually gets set unconditionally when the swapchain is created. It's set to `NO` if you passed `VK_PRESENT_MODE_IMMEDIATE_KHR` as the `presentMode` in the `pCreateInfo`. The `presentInfo.presentMode` is only used in the `VK_EXT_swapchain_maintenance1` case, if you passed a `VkSwapchainPresentModeInfoEXT` struct with your presentation command.

**Comment by billhollings** [comment]
> The following code inside `MVKPresentableSwapchainImage::presentCAMetalDrawable()` fixes it. However, I'm not sure if this is how you want to solve it. For instance, it appears the meaning of `presentInfo.hasPresentTime` may be overloaded, and currently indicates that both a presentedHandler and desiredPresentTime are requested. In my case I am only interested in the presented time handler, and my desired present time is not defined (i.e. = 0).

Once again, thanks for your research on this.

The Metal docs for [`presentAtTime:`](https://developer.apple.com/documentation/metal/mtldrawable/1470282-presentattime?language=objc) says...

> `presentationTime` The Mach absolute time at which the drawable should be presented, in seconds.
> ...
> If the commands complete after the presentation time, the drawable presents its contents as soon as possible.

And Mach absolute time is not strictly defined, but generally it is the uptime since last reboot (documented elsewhere). So, we should expect a `presentationTime ` of zero would describe some time far in the relative past, for which the commands definitely complete after, and the drawable should present immediately.

So one way or another, Apple is definitely not doing what it advertises.

Thanks for the suggested fix. Looks like it makes sense, and I'll include a variation of it.

But before I do, I wonder if zero has a special meaning inside Apple's implementation of `presentAtTime:`. My concern is what happens when `presentAtTime:` is not zero, but is still in the past.

Are you able to test this once more with a `desiredPresentTime` that is not zero, but is definitely in the past (ie. set `desiredPresentTime` to current Mach time when you submit it)? Doing so will trigger your first code path, and it will be interesting to see if the Metal operation does what it says, and presents immediately.

If it doesn't, we may have to expand your `if (presentInfo.desiredPresentTime)` to compare agains current Mach time instead of just zero.

**Comment by SRSaunders** [comment]
As you requested, I tested this with `desiredPresentTime` set to `clock_gettime( CLOCK_MONOTONIC, &ts )`.  Here is a screenshot from Xcode showing the `presentTiming` values returned by `vkGetPastPresentationTimingGOOGLE()`:

![Screen Shot 2023-06-01 at 9 22 32 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/06280907-a4c4-4f61-9cdc-8daef6df0914)
As you can see, `desiredPresentTime` (6964.718 sec) is about .47 sec earlier than `actualPresentTime` (6965.188 sec), or the equivalent of 28 frames @ 60 Hz.  And yes, it seems that Metal is vsyncing when `[mtlDrwbl presentAtTime: ...]` is used vs. `[mtlDrwbl present]`.  Do the Metal docs for `presentAtTime` imply or rule out vsyncing?  For instance, what do the words "as soon as possible" really mean - i.e. immediately, or wait until vsync?   It seems a big ambiguous to me.  In any case, perhaps the case when `desiredPresentTime` is non-zero should not be of great concern for this particular defect.  

From the Vulkan docs regarding `desiredPresentTime`:

> desiredPresentTime specifies that the image given should not be displayed to the user any earlier than this time... A value of zero specifies that the presentation engine may display the image at any time. This is useful when the application desires to provide presentID

This is my specific case, and why I suggested special handling (i.e. `[mtlDrwbl present]`) when `desiredPresentTime = 0`.  And Vulkan seems to give some room for interpretation when a non-zero value is provided.

Perhaps the `desiredPresentTime = 0` case is all we have to solve for the moment.  Your thoughts?

**Comment by SRSaunders** [comment]
Thanks for all the time you are spending thinking about this.  However, I have concerns about modifying the default semantics too much.  I believe the `VK_GOOGLE_display_timing` feature was originally put in place for Android, where the only non-shared presentation modes are `VK_PRESENT_MODE_FIFO_KHR` and `VK_PRESENT_MODE_MAILBOX_KHR`, both which are vsync-on modes.  `VK_PRESENT_MODE_IMMEDIATE_KHR` is not available on that mobile platform for power-saving reasons.  MoltenVK's implementation of the feature is a bit of a hybrid, as it brings the capability to the Apple environment where you have to adapt to Metal semantics.  It is not available with Vulkan on Windows and Linux.

I am not sure what Apple's Metal design goals were for `presentAtTime`, whether this was to mirror the Android feature for mobile/iOS, or for desktop/macOS as well.  Ironically, I believe the original Vulkan Android semantics designed for vsync-on happen to line up with Metal's semantics for `presentAtTime`, also with vsync-on.  I am speculating, but perhaps presentation control for mobile power management has been the goal all along for both platforms.  Apple's Metal docs may not be clear about vsync interaction, but the Vulkan `VK_GOOGLE_display_timing` feature description is very clear that the return fields are meant for presentation control in the context of vsync.  The example use cases are for retarding or accelerating presentation rate with refresh duration granularity (i.e. not fine-grained), and/or for minimizing render-to-present latency in the context of vsync on.

There is a more general, but work-in-progress extension called `VK_EXT_present_timing`, which has been contemplated for a long time.  Work is still ongoing - see https://github.com/KhronosGroup/Vulkan-Docs/pull/1364.  I am pretty sure that effort is trying to handle all cases - mobile and desktop cross-platform, and that is the reason for a lengthy dev period.

In the end, I suspect we are in uncharted territory here and I am asking for a special case with partial support of this Android feature when using the `VK_PRESENT_MODE_IMMEDIATE_KHR` presentation mode.  Original Vulkan semantics were for vsync-on presentation, and I am asking for special vsync-off handling when `desiredPresentTime = 0`.

In summary:

1. I only need to handle the case of `VK_PRESENT_MODE_IMMEDIATE_KHR`/vsync-off when `desiredPresentTime = 0`
2. Other devs may expect vsync-on Vulkan Android semantics to carry over when `desiredPresentTime != 0` (however with limitations since `earliestPresentTime` and `presentMargin` are not available with Metal)
3. Since the additional code you suggest is inline with the presentation loop, I would be concerned about adding a mandatory system call (for sys time) with comparison operations on every cycle.  What would be the performance impact of this for `VK_PRESENT_MODE_IMMEDIATE_KHR` where FPS performance and low latency is the goal?

You may be looking at implementing `VK_EXT_present_timing` sometime in the future, and perhaps leaving the more general discussion and effort to then might be better.  Just my 2 cents...

**Comment by billhollings** [comment]
PR #1936 fixes this issue. Please retest with latest MoltenVK and close this issue if it resolves your problem.

I've implemented your design. Thanks for providing it, and for all the design discussion input. I am a bit concerned that we are interpreting the Vulkan spec on this in a way that was not intended, since we are special-casing behavior based on one value, without it being defined in the spec. But since the spec is not explicit about what `desiredPresentTime = 0` means, this is probably as close as we can get to meshing the difference between Vulkan and Metal present timing behavior. And I agree that adding an additional test for current time would slow down what is meant to be a fast-path.

> `VK_PRESENT_MODE_IMMEDIATE_KHR` is not available on that mobile platform for power-saving reasons.

It's not available on the iOS mobile platform, either, and probably for the same reasons.

**Comment by SRSaunders** [comment]
Thanks very much. I like how you have separated and sequentialized the test logic for `presentInfo.hasPresentTime` and `presentInfo.desiredPresentTime` - very smart.  If `presentInfo.hasPresentTime` is false, then by implication and initialization `presentInfo.desiredPresentTime` will also be zero.

I have tested this and it resolves my issue.  Thanks again.
